### ISSUE TITLE
Make the extension work with Reddit's redesign

### ISSUE THREAD

**Issue by pkrakowiak** [easyfuckingpeasy/chrome-reddit-comment-highlights#18]
Reddit has a new design (as seen in https://www.reddit.com/r/redesign/) which is being rolled out (you can also opt-in right now) to users. This extension no longer works (well, the markup changed...). Is there any chance it will be updated to support the new website design?

**Comment by aesy** [comment]
Hey, thanks for mentioning that! That's something that will definitely be fixed, I will take a look at it sometime this weekend.

**Comment by aesy** [comment]
This was a lot harder than I anticipated, some help would be greatly appreciated.

I'm thinking aloud here to structure my thoughts.

My struggles so far:

Struggle: How to detect page changes?
Best solution so far: Listen for `onpopstate` events and monkeypatch `history.pushState` and `history.replaceState`. Patching the environment like that doesn't work from a webextension content script, but that can be circumvented by injecting a script that patches the environment and sends events which the content script then listens to... Hacky as hell.

Struggle: How to add styles to specific comments?
Best solution so far:  Many elements are missing classnames though, so the comment styling will be bound to their structure in the DOM. It's tedious to style and sensitive to change. Also, when adding a classname to a react component, it is removed the next time it's re-rendered... I have no idea how to work around that.

Struggle: How to get the ids of all comments on the page?
Best solution so far: Comments fortunately have a `Comment`-classname, and they are tagged with their respective ids, so it's possible to query all comments and read the ids. Using a `MutationObserver` to check for changes, like done previously, should still work.

Struggle: How to get all the children of the comments?
Best solution so far: Comment threads are not structured in the DOM the same way they are logically related, they are ordered as a flat list. That makes it tedious to get all the children of a comment, but it's possible with the help of the threadlines which are tagged with the id of the comment they belong to.

```js
const childComments = [...document.getElementsByClassName("<id-of-comment>")]
	.map(e => e.parentElement)
	.map(e => e.nextSibling)
	.filter(e => !!e)
	.filter(e => e.classList.contains("Comment"))
``` 

Struggle: How to inject additional components?
Best solution so far: None.

There will be many more struggles to get the extension running like before, and most of these solutions feel like quickfixes. I would've preferred to hook into the React application and read from and manipulate the redux store directly. I would've also preferred to do it in some way that isn't using `__REACT_DEVTOOLS_GLOBAL_HOOK__` as this could (I wouldn't be surprised) be turned off any time.

I would appreciate any response. How does RES handle this?

**Comment by aesy** [comment]
I'll keep using this page as my scratchpad :)

Struggle: How to get the currently logged in user now that all the classnames are gone?
Best solution so far: 
```js
fetch("https://www.reddit.com/api/me.json", { credentials: "same-origin" })
	.then(r => r.json())
	.then(r => console.log(r.data.name));
```

Struggle: How to know if the page is using the redesign or old design?
Best solution so far: The API endpoint above includes a property called `in_redesign_beta`. If it's false or undefined, assume old site? How reliable is that?

I'm starting to give up to be honest.

**Comment by aesy** [comment]
I have a lot on my plate right now, but I'll give this another shot at a later time ~approx late June. Pull requests are still very welcome though.

**Comment by Dangelman** [comment]
Is there any update on this? Currently using old, but I'd really like to switch to redesign and still use the extension.

It's such a pity Reddit locked such an useful feature behind a subscription (then again, it's the only useful thing the subscription offers beside adblock).

**Comment by aesy** [comment]
Sorry to say but I don't have much motivation to finish this as I find the "new" site to be pretty awful and will continue to use the old site exclusively.

**Comment by aesy** [comment]
Implemented in 7801b3a383a9f0295dc820cc04994a8b9d86fe2d.

I did the thing... I was about to close this issue as 'wont do', but since most of the work was already done, I instead took some time to finish it. It seems to work well enough, but it's possible that it's semi-broken due to the nature of how React works. If that's the case then I might remove it in the future.

I will publish a new release within a week or two.
